### ISSUE TITLE
feature: having different vlan pool allocations for UNIs and NNIs

### ISSUE THREAD

**Issue by viniarck** [kytos-ng/mef_eline#545]
Currently, `available_tags` belongs to the interface, but currently, it's supported that an EVC can also be created in a NNI interface, which starts acting as a UNI for that EVC. In this case, since it uses the same pool, link down events for the NNI being used by other EVCs might allocate the UNI s vlan if that ever gets released by the EVC (by changing it or deleting), so changing certain configurations might cause this dispute that can be cumbersome for network operators. 

Ultimately, it's desired to have an explicit non overlapping pool for UNIs and NNIs of a given interface. That will be discussed when this gets prioritized, making sure if meets the current usages that well known in production. We're not too far away from this, it's mostly a matter of having a separate pool while still allocating.

Related prior discussions: https://github.com/kytos-ng/kytos/discussions/406

**Comment by Ktmi** [comment]
I've been writing a some documents regarding this and other issues, here's a bit of it. What do you think @viniarck?

# Rethinking How We Define the Topology

Some of our terminology is a bit confusing with how we use it, considering the context that we are trying to resolve. Part of that is the rigidity that the terminology suggests. Network-to-Network interface(NNI) and User-to-Network Interface (UNI), which both suggest a physical interface connecting to a single destination. However, we know that this is not always the case, so let's take some steps to redefine a few terms.

First, we'll be considering all of this terminology and how its used in the unidirectional case, unless otherwise specified. It's simpler to explain, and ultimately bidirectional network traffic flow is just an abstraction for a pair of unidirectional network traffic flows.

Second, we will be replacing the terms UNI and NNI. They don't really give us the full picture on what's going on. Instead we are going to be using the following terminology:

- A switch has a set of **Ingress** and **Egress** rules. Some Ingress rules have an expected source, and some egress rules have an expected destination.
- When you have an egress rule which we can reasonably expect its traffic to received by a particular ingress rule, you then have a **Transport**.
- From an ingress, egress, or transport, we can select a subset of the traffic going through them through a **Slice**. These slices are just another ingress/egress/transport, but with additional parameters on the expected traffic.
- When we want to route packets match by a particular ingress rule to an egress rule, we must define a **Connection** between the two.

**Comment by viniarck** [comment]
cc'ing @Alopalao in the discussion too if he wants to chime in too since he also worked in this part before.

**Comment by Alopalao** [comment]
| Split the current pool into two disjoint pools for UNI and NNI

Does this solution implies to replace `available_tags` and `tag_ranges` like this?:
```
        self.available_tags = {
            "nni": {'vlan': self.default_tag_values['vlan']},
            "uni": {'vlan': self.default_tag_values['vlan']},
        } 
        self.tag_ranges = {
            "nni": {'vlan': self.default_tag_values['vlan']},
            "uni": {'vlan': self.default_tag_values['vlan']}
        }
```
If yes, we could add parameters to the functions involved in modifying these lists. Maybe `role_type`, in the same way we pass around `tag_type` in the methods. `role_type` expected values would be `"nni"` and `"uni"`.
With default parameters as `role_type="nni"`, and a fast reading for anything related, looks like not too much would be broken.
- Changes are needed in `kytos/core/interface.py`.
- DB `interface_details` will need to be updated. Also taking care of EVC UNIs.
- Change `mef_eline` so UNI can start adding `role_type="uni"` to anything related to TAGs

An extra problem that I found is that we could be having duplicated EVCs that do not have TAG, more in [issue](https://github.com/kytos-ng/mef_eline/issues/569).

**Comment by Ktmi** [comment]
The thing is that when we take out a slice of a physical interface for routing packets, we aren't really taking a slice out of the interface, but of the transport. The transport should have already reserved the slice of the physical interface it is using. That's why I went with the more hierarchical design.

How about this, interfaces have some basic reservations on them representing either links or unis. And then further slicing of the links is tied to the link specifically. It would look something like this:

```python
interface.tag_range = [[1, 4095]]
interface.available_tags = [[51, 99], [501, 4095]]
interface.reservations = [
    {
	    "owner": "mef_eline",
	    "labels": {"evc-247", "endpoint_a"},
	    "tag": {"vlan": {"value": [1, 50]}},
	},
    {
	    "owner": "topology",
	    "labels": {"link", "link-42"},
	    "tag": {"vlan": {"value": [100, 500]},
	}, 
]
link.reservations = [
	{
		"owner": "mef_eline",
		"labels": {"evc-248", "transport"},
		"tag": {
			"vlan": {"value": 100},
		}
	}
]
link.tag_range = [[100, 500]]
link.available_tags = [[101, 500]]
```

**Comment by viniarck** [comment]
> If yes, we could add parameters to the functions involved in modifying these lists. Maybe role_type, in the same way we pass around tag_type in the methods. role_type expected values would be "nni" and "uni".

Yes @Alopalao. The biggest problem with this though, is that `tag_ranges` should be disjoint too, and therefore which default `tag_ranges` should be set? [The default values](https://github.com/kytos-ng/kytos/blob/master/kytos/core/interface.py#L229), was safe up to this point since we have only a single pool for the interface. Since we never know upfront if the interface will be used as a UNI, how should be pool be divided? We could still use the default `[[1, 4095]]` for UNIs and `[[]]` NNIs, and then set `[[]]` UNI and `[[1, 4095]]` when a link is discovered, and then from that point forward network operators would need to configure the UNI tag_ranges accordingly for the values they expect to use.

Thinking about it more, having all explicit created pools might be too much work and inconvenience, and refactoring just to support the UNI as NNI case. With, that in mind, that's why I started thinking about potentially borrowing/slicing and creating dynamic pools from a given interface, since at that point we could defer entirely for the network operator to create with the tag_ranges without having to change much.

```
interface.tag_ranges =  {'vlan': [[1, 4095]]}
interface.special_tags = {'vlan': ['untagged', 'any']}
interface.available_tags {'vlan': [[1, 3798], [3800, 4095]]}

interface.sub_tag_ranges = {'<sub_pool_name>': 'vlan': [[1, 100]]}  # this should be within interface.tag_ranges, and configurd by a network operator, when this is configured, it would borrow from interface.available_tags
interface.sub_available_tags = {'<sub_pool_name>': 'vlan': [[1, 100]]}

# after the borrow  interface.available_tags would be, so none of the sub_available_tags would be available in the general pool anymore:
interface.available_tags {'vlan': [[101, 3798], [3800, 4095]]}
```
The biggest drawback is that this would require the disclipline of knowing from which pool the application would be allocating, which so far we don't have many corner cases, so it would be only when mef_eline would be allocating it's UNI, if it's a NNI interface, it would try to allocate from a sub pool. Essentially, it's not too different from the disjoint solution, except is dynamic created, and the main tag_ranges behaves like a super set. Anyway, just an initial ideia to bring to the table here.

**Comment by viniarck** [comment]
@Ktmi, 

> how about this, interfaces have some basic reservations on them representing either links or unis. And then further slicing of the links is tied to the link specifically. It would look something like this:

a `Link` having a dedicated `tag_ranges` and `available_tags` borrowed from the interface `available_tags` seems interesting, as long as we keep it coherent and only allow a valid sub ``tag_ranges`` to be configured, and by default it should have the same tag_ranges as the underlying interfaces. Seems like a subpool deriving from the interface except it'd be a bit more ergonomic and direct on a `Link`. In the future to support other kinds of interfaces it would be difficult, but that's not on our radar in the short/medium term so it can be a compromise that we can make. 

Points of attention that I'd like to see:

- We'll need to also consider how it'll be stored in the DB and how it'll coexist with `interface_details`
- The reservations are good, although they're not a must have part of the sollution, but certainly desirable. I can see this being helpful for correlating and general debug too. Backfilling prod documents won't be too easy though, but can be done with scripts.
- Endpoints to configure link tag_ranges
- DB scripts for prod

Keep refining the solution and the rest of the proposal, @Ktmi, great idea, it has good potential, and let's weigh with the other options and then we'll settle on an option and we'll maintain it. Thanks @Alopalao for your contributions here too.

**Comment by viniarck** [comment]
Regarding the reservations, let's deliver it incrementally and leave it to be refined and further discussed in the future, especially since this needs to be backported.

The tradeoff with reservations we'd be gaining valuable data about reservations but paying the cost with addtional writes, plus we'd need efficient indexes when also deleting the documents, which certainly is doable but won't be trivial, and having to discuss all of that for a backport will be too much. So, let's go incrementally here and not lose this idea for the future, it has great potential. OK? 

Other than that, when the rest of the discussion proposal is ready ping me and I can do a final review.

**Comment by viniarck** [comment]
Removing label `2025.2`, and setting `2026.1` as we spoke on today's meeting. 

This might land as a patch on 2025.2 later
